# Thread 2: Gender Injustice

### This thread looks at the crimes against women in the past decade and concludes important pointers from it. For ex- how the rate of crime against women is affected each year, whether criminals are convicted, etc.

## EDA 

# 01. Importing libraries

In [1]:
import pandas as pd
import sys
sys.path.append("..")

import utility as u

# 02. EDA on each file before using it for analysis

## a) Crime against women (2001-2022).csv

In [2]:
# this file uses wide format, we will first convert it into long format

df= u.load_data("data/Crime against Women (2001-2022).csv")
csv1= u.wide_to_long(df=df, id_cols="CRIME HEAD", value_name="cases", var_name="year")

In [3]:
u.eda(df=csv1)

  DataFrame
Shape       : 242 rows × 3 cols

Dtypes:
CRIME HEAD    object
year          object
cases          int64

Null counts:
CRIME HEAD    0
year          0
cases         0

Null %:
CRIME HEAD    0.0
year          0.0
cases         0.0

Sample:
                                  CRIME HEAD  year  cases
0                                       Rape  2001  16075
1  Kidnapping and Abduction of Women & Girls  2001  14645
2                               Dowry Deaths  2001   6851


In [4]:
u.missing_report(df=csv1)

  DataFrame
Shape       : 242 rows × 3 cols

             dtype  null_count  null_% flag
CRIME HEAD  object           0     0.0     
year        object           0     0.0     
cases        int64           0     0.0     

No columns exceed 50% null threshold.


## Observations from csv1-
--> No nulls in the dataset.

--> Wide format converted to long: 11 crime heads × 22 years = 242 rows, 3 cols.

--> No nulls anywhere, clean dataset.

--> `year` column is object dtype after melt — needs cast to int in cleaning.

--> National level only, no state breakdown.

## b) combined_crime_against_women_full.csv

In [5]:
csv2= u.load_data("data/combined_crime_against_women_full.csv")

In [6]:
csv2.head

<bound method NDFrame.head of       Year           State_UT   State_UT_Raw   Row_Type  \
0     2001  All India (Total)      All India  Aggregate   
1     2001  All India (Total)      All India  Aggregate   
2     2001  All India (Total)      ALL INDIA  Aggregate   
3     2001  All India (Total)      All India  Aggregate   
4     2001  All India (Total)      All India  Aggregate   
...    ...                ...            ...        ...   
6328  2025          Telangana      Telangana   State/UT   
6329  2025            Tripura        Tripura   State/UT   
6330  2025      Uttar Pradesh  Uttar Pradesh   State/UT   
6331  2025        Uttarakhand    Uttarakhand   State/UT   
6332  2025        West Bengal    West Bengal   State/UT   

                                Crime_Head_Standardized  \
0       Assault on Women with Intent to Outrage Modesty   
1                       Cruelty by Husband or Relatives   
2                                          Dowry Deaths   
3                      Im

In [7]:
u.eda(csv2)

  DataFrame
Shape       : 6333 rows × 8 cols

Dtypes:
Year                         int64
State_UT                    object
State_UT_Raw                object
Row_Type                    object
Crime_Head_Standardized     object
Crime_Head_Raw              object
Is_Detailed_Subcategory       bool
Count                      float64

Null counts:
Year                         0
State_UT                     0
State_UT_Raw                 0
Row_Type                     0
Crime_Head_Standardized    975
Crime_Head_Raw               0
Is_Detailed_Subcategory      0
Count                      116

Null %:
Year                        0.00
State_UT                    0.00
State_UT_Raw                0.00
Row_Type                    0.00
Crime_Head_Standardized    15.40
Crime_Head_Raw              0.00
Is_Detailed_Subcategory     0.00
Count                       1.83

Sample:
   Year           State_UT State_UT_Raw   Row_Type                          Crime_Head_Standardized                       

In [8]:
u.missing_report(csv2)

  DataFrame
Shape       : 6333 rows × 8 cols

                           dtype  null_count  null_% flag
Year                       int64           0    0.00     
State_UT                  object           0    0.00     
State_UT_Raw              object           0    0.00     
Row_Type                  object           0    0.00     
Crime_Head_Standardized   object         975   15.40     
Crime_Head_Raw            object           0    0.00     
Is_Detailed_Subcategory     bool           0    0.00     
Count                    float64         116    1.83     

No columns exceed 50% null threshold.


In [9]:
csv2 = csv2[csv2["Row_Type"] == "State/UT"].reset_index(drop=True)
csv2.head(10)

,Year,State_UT,State_UT_Raw,Row_Type,Crime_Head_Standardized,Crime_Head_Raw,Is_Detailed_Subcategory,Count
0,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Assault on Women with Intent to Outrage Modesty,ASSAULT ON WOMEN WITH INTENT TO OUTRAGE HER MO...,False,19.0
1,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Cruelty by Husband or Relatives,CRUELTY BY HUSBAND OR RELATIVES,False,9.0
2,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Dowry Deaths,DOWRY DEATH,False,0.0
3,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Immoral Traffic (Prevention) Act,IMMORAL TRAFFIC(PREVENTION)ACT,False,0.0
4,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Indecent Representation of Women (Prevention) Act,INDECENT REPRESENTATION OF WOMEN(PREVENTION)ACT,False,0.0
5,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Insult to Modesty of Women,INSULT TO THE MODESTY OF WOMEN,False,1.0
6,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Kidnapping & Abduction of Women,KIDNAPPING & ABDUCTION,False,2.0
7,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Rape,RAPE,False,3.0
8,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Total Crimes against Women,TOTAL CRIMES AGAINST WOMEN,False,34.0
9,2001,Andhra Pradesh,Andhra Pradesh,State/UT,Assault on Women with Intent to Outrage Modesty,ASSAULT ON WOMEN WITH INTENT TO OUTRAGE HER MO...,False,3544.0


In [10]:
csv2["State_UT"].unique()

array(['Andaman & Nicobar Islands', 'Andhra Pradesh', 'Arunachal Pradesh',
       'Assam', 'Bihar', 'Chandigarh', 'Chhattisgarh',
       'Dadra & Nagar Haveli (pre-2020, separate UT)',
       'Daman & Diu (pre-2020, separate UT)', 'Delhi', 'Goa', 'Gujarat',
       'Haryana', 'Himachal Pradesh', 'Jammu & Kashmir', 'Jharkhand',
       'Karnataka', 'Kerala', 'Lakshadweep', 'Madhya Pradesh',
       'Maharashtra', 'Manipur', 'Meghalaya', 'Mizoram', 'Nagaland',
       'Odisha', 'Puducherry', 'Punjab', 'Rajasthan', 'Sikkim',
       'Tamil Nadu', 'Tripura', 'Uttar Pradesh', 'Uttarakhand',
       'West Bengal', 'Telangana', 'Dadra & Nagar Haveli and Daman & Diu',
       'Ladakh'], dtype=object)

Dadra & Nagar Haveli and Daman & Diu appear as separate entities pre-2020 and merged post-2020 due to the 2020 reorganisation. Will be handled in cleaning.

In [11]:
print(csv2["Year"].unique())
print(csv2["Year"].min(), "-", csv2["Year"].max())

[2001 2002 2003 2004 2005 2006 2007 2008 2009 2010 2011 2012 2013 2014
 2015 2016 2017 2018 2019 2020 2021 2022 2023 2024 2025]
2001 - 2025


The datasets looks at the crimes committed across more than a decade. 

From 2001-2025

In [12]:
csv2[csv2["Crime_Head_Standardized"].isna()][["Year","State_UT","Crime_Head_Raw"]].drop_duplicates().head(15)

,Year,State_UT,Crime_Head_Raw
4201,2014,Andaman & Nicobar Islands,1.1 - Custodial Rape
4202,2014,Andaman & Nicobar Islands,1.1.1 - Custodial_Gang Rape
4203,2014,Andaman & Nicobar Islands,1.1.2 - Custodial_Other Rape
4204,2014,Andaman & Nicobar Islands,1.2 - Rape other than Custodial
4205,2014,Andaman & Nicobar Islands,1.2.1 - Rape_Gang Rape
4206,2014,Andaman & Nicobar Islands,1.2.2 - Rape_Others
4212,2014,Andaman & Nicobar Islands,14.1 - Under ITP Section 5
4213,2014,Andaman & Nicobar Islands,14.2 - Under ITP Section 6
4214,2014,Andaman & Nicobar Islands,14.3 - Under ITP Section 7
4215,2014,Andaman & Nicobar Islands,14.4 - Under ITP Section 8


In [13]:
csv2[csv2["Count"].isna()][["Year","State_UT","Crime_Head_Standardized"]].head(20)

,Year,State_UT,Crime_Head_Standardized
5657,2015,Ladakh,Total Crimes against Women
5693,2016,Ladakh,Total Crimes against Women
5729,2017,Ladakh,Total Crimes against Women
5765,2018,Ladakh,Total Crimes against Women
5801,2019,Ladakh,Total Crimes against Women
5928,2023,Andaman & Nicobar Islands,Total Crimes against Women
5929,2023,Andhra Pradesh,Total Crimes against Women
5930,2023,Arunachal Pradesh,Total Crimes against Women
5931,2023,Assam,Total Crimes against Women
5932,2023,Bihar,Total Crimes against Women


In [14]:
csv2.groupby("Year")["Count"].apply(lambda x: x.isna().sum())

Year
2001     0
2002     0
2003     0
2004     0
2005     0
2006     0
2007     0
2008     0
2009     0
2010     0
2011     0
2012     0
2013     0
2014     0
2015     1
2016     1
2017     1
2018     1
2019     1
2020     0
2021     0
2022     0
2023    31
2024    36
2025    36
Name: Count, dtype: int64

In [15]:

# confirm Is_Detailed_Subcategory accounts for the standardized nulls
csv2[csv2["Is_Detailed_Subcategory"] == True].shape  

(900, 8)

In [16]:
csv2["Crime_Head_Standardized"].dropna().unique()

array(['Assault on Women with Intent to Outrage Modesty',
       'Cruelty by Husband or Relatives', 'Dowry Deaths',
       'Immoral Traffic (Prevention) Act',
       'Indecent Representation of Women (Prevention) Act',
       'Insult to Modesty of Women', 'Kidnapping & Abduction of Women',
       'Rape', 'Total Crimes against Women',
       'Commission of Sati (Prevention) Act', 'Dowry Prohibition Act',
       'Importation of Girls from Foreign Country',
       'Protection of Women from Domestic Violence Act',
       'Attempt to Commit Rape', 'Abetment of Suicide of Women'],
      dtype=object)

## Observations from csv2
--> 6333 rows × 8 cols, covers 2001–2025

--> Filtered to `Row_Type == "State/UT"` to remove aggregates

--> 38 unique states/UTs

--> `Crime_Head_Standardized` has 975 nulls (15.4%) — all correspond to  `Is_Detailed_Subcategory == True` rows (granular subcategories like "1.1 - Custodial Rape", "3.2 - Kidnapping in order to Murder" etc.) that don't have a standardized label. Not data loss, just unmapped subcategories — drop or handle separately in cleaning

--> `Count` has 116 nulls (1.83%) — concentrated in 2023 (31), 2024 (36), 2025 (36) and Ladakh 2015–2019 (5). Recent years likely not fully reported yet; Ladakh pre-2020 is expected since it wasn't a separate UT then

--> 15 standardized crime heads including newer ones like `Protection of Women from Domestic Violence Act`, `Attempt to Commit Rape`, `Abetment of Suicide of Women` — not present in csv1

--> Dadra & Nagar Haveli and Daman & Diu: separate pre-2020, merged post-2020 — needs handling in cleaning

## c) NCRB_CII_2023_Table_3A.8_0.csv

In [17]:
csv3= u.load_data("data/NCRB_CII_2023_Table_3A.8_0.csv")
csv3.head(10)

,Sl. No.,State/UT,Cases Pending Trial from the Previous Year - ( Col. 3),Cases Sent for Trial during the year - ( Col. 4),Total Cases for Trial (Col.3+Col.4) - ( Col. 5),Cases Abated by Court - ( Col. 6),Cases Withdrawn from Prosecution - ( Col. 7),Cases Compounded or Compromised - ( Col. 8),Cases Disposed off by Plea Bargaining - ( Col. 9),Cases Quashed - ( Col. 10),...,Cases Convicted Out of Cases from Previous Year - ( Col. 13),Cases Convicted Out of Cases during the Year - ( Col. 14),Cases Convicted (Col.13+Col.14) - ( Col. 15),Cases Discharged - ( Col. 16),Cases Acquitted - ( Col. 17),Cases in which Trials were Completed (Col.15+Col.16+ Col.17) - ( Col. 18),Cases Disposed off by Courts (Col.11+Col.18) - ( Col. 19),Cases Pending Trial at End of the Year (Col.5-Col.19) - ( Col. 20),Conviction Rate (Col.15/Col.18)*100 - ( Col. 21),Pendency Percentage (Col.20/Col.5)*100 - ( Col. 22)
0,1,Andhra Pradesh,59575,21180,80755,152,6,7340,0,33,...,466,87,553,142,8855,9550,17081,63674,5.8,78.8
1,2,Arunachal Pradesh,3137,255,3392,3,0,20,1,3,...,23,1,24,8,38,70,97,3295,34.3,97.1
2,3,Assam,90075,11192,101267,227,5958,6355,1,0,...,467,265,732,36,9066,9834,22375,78892,7.4,77.9
3,4,Bihar,115796,17370,133166,0,0,338,0,0,...,1114,171,1285,65,932,2282,2620,130546,56.3,98.0
4,5,Chhattisgarh,34270,5931,40201,65,0,70,95,0,...,773,191,964,625,2777,4366,4596,35605,22.1,88.6
5,6,Goa,1811,204,2015,6,0,11,0,0,...,13,1,14,9,111,134,151,1864,10.4,92.5
6,7,Gujarat,98357,6855,105212,15,9,23,0,52,...,425,51,476,180,6531,7187,7286,97926,6.6,93.1
7,8,Haryana,41260,8987,50247,9,0,0,0,218,...,779,39,818,29,5159,6006,6233,44014,13.6,87.6
8,9,Himachal Pradesh,9370,1146,10516,3,63,25,0,111,...,135,1,136,5,707,848,1050,9466,16.0,90.0
9,10,Jharkhand,30019,6353,36372,30,1,50,1,0,...,1215,469,1684,29,3781,5494,5576,30796,30.7,84.7


In [18]:
u.eda(csv3)

  DataFrame
Shape       : 39 rows × 22 cols

Dtypes:
Sl. No.                                                                            object
State/UT                                                                           object
Cases Pending Trial from the Previous Year - ( Col. 3)                              int64
Cases Sent for Trial during the year - ( Col. 4)                                    int64
Total Cases for Trial (Col.3+Col.4) - ( Col. 5)                                     int64
Cases Abated by Court - ( Col. 6)                                                   int64
Cases Withdrawn from Prosecution - ( Col. 7)                                        int64
Cases Compounded or Compromised - ( Col. 8)                                         int64
Cases Disposed off by Plea Bargaining - ( Col. 9)                                   int64
Cases Quashed - ( Col. 10)                                                          int64
Cases Disposed off without trial (Col.6+Col.7+C

In [19]:
u.missing_report(csv3)

  DataFrame
Shape       : 39 rows × 22 cols

                                                                                  dtype  null_count  null_% flag
Sl. No.                                                                          object           0     0.0     
State/UT                                                                         object           0     0.0     
Cases Pending Trial from the Previous Year - ( Col. 3)                            int64           0     0.0     
Cases Sent for Trial during the year - ( Col. 4)                                  int64           0     0.0     
Total Cases for Trial (Col.3+Col.4) - ( Col. 5)                                   int64           0     0.0     
Cases Abated by Court - ( Col. 6)                                                 int64           0     0.0     
Cases Withdrawn from Prosecution - ( Col. 7)                                      int64           0     0.0     
Cases Compounded or Compromised - ( Col. 8)        

In [20]:
csv3= u.drop_totals(csv3, "State/UT")
csv3.head(30)

Dropped 3 aggregate rows from 'State/UT'


,Sl. No.,State/UT,Cases Pending Trial from the Previous Year - ( Col. 3),Cases Sent for Trial during the year - ( Col. 4),Total Cases for Trial (Col.3+Col.4) - ( Col. 5),Cases Abated by Court - ( Col. 6),Cases Withdrawn from Prosecution - ( Col. 7),Cases Compounded or Compromised - ( Col. 8),Cases Disposed off by Plea Bargaining - ( Col. 9),Cases Quashed - ( Col. 10),...,Cases Convicted Out of Cases from Previous Year - ( Col. 13),Cases Convicted Out of Cases during the Year - ( Col. 14),Cases Convicted (Col.13+Col.14) - ( Col. 15),Cases Discharged - ( Col. 16),Cases Acquitted - ( Col. 17),Cases in which Trials were Completed (Col.15+Col.16+ Col.17) - ( Col. 18),Cases Disposed off by Courts (Col.11+Col.18) - ( Col. 19),Cases Pending Trial at End of the Year (Col.5-Col.19) - ( Col. 20),Conviction Rate (Col.15/Col.18)*100 - ( Col. 21),Pendency Percentage (Col.20/Col.5)*100 - ( Col. 22)
0,1,Andhra Pradesh,59575,21180,80755,152,6,7340,0,33,...,466,87,553,142,8855,9550,17081,63674,5.8,78.8
1,2,Arunachal Pradesh,3137,255,3392,3,0,20,1,3,...,23,1,24,8,38,70,97,3295,34.3,97.1
2,3,Assam,90075,11192,101267,227,5958,6355,1,0,...,467,265,732,36,9066,9834,22375,78892,7.4,77.9
3,4,Bihar,115796,17370,133166,0,0,338,0,0,...,1114,171,1285,65,932,2282,2620,130546,56.3,98.0
4,5,Chhattisgarh,34270,5931,40201,65,0,70,95,0,...,773,191,964,625,2777,4366,4596,35605,22.1,88.6
5,6,Goa,1811,204,2015,6,0,11,0,0,...,13,1,14,9,111,134,151,1864,10.4,92.5
6,7,Gujarat,98357,6855,105212,15,9,23,0,52,...,425,51,476,180,6531,7187,7286,97926,6.6,93.1
7,8,Haryana,41260,8987,50247,9,0,0,0,218,...,779,39,818,29,5159,6006,6233,44014,13.6,87.6
8,9,Himachal Pradesh,9370,1146,10516,3,63,25,0,111,...,135,1,136,5,707,848,1050,9466,16.0,90.0
9,10,Jharkhand,30019,6353,36372,30,1,50,1,0,...,1215,469,1684,29,3781,5494,5576,30796,30.7,84.7
